src/VisualStudio/Core/Def/Implementation/DebuggerIntelliSense/DebuggerIntellisenseFilter.cs

```diff
@@ -16,26 +17,63 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.DebuggerIntelliSense
 {
-    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable
+    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable, IFeatureController
         where TPackage : AbstractPackage<TPackage, TLanguageService>
         where TLanguageService : AbstractLanguageService<TPackage, TLanguageService>
     {
         private readonly ICommandHandlerServiceFactory _commandFactory;
         private readonly IWpfTextView _wpfTextView;
+        private readonly IFeatureServiceFactory _featureServiceFactory;
         private AbstractDebuggerIntelliSenseContext _context;
         private IOleCommandTarget _originalNextCommandFilter;
+        private IFeatureDisableToken _completionDisabledToken;
 
         internal bool Enabled { get; set; }
 
         public DebuggerIntelliSenseFilter(
             AbstractLanguageService<TPackage, TLanguageService> languageService,
             IWpfTextView wpfTextView,
             IVsEditorAdaptersFactoryService adapterFactory,
-            ICommandHandlerServiceFactory commandFactory)
+            ICommandHandlerServiceFactory commandFactory,
+            IFeatureServiceFactory featureServiceFactory)
             : base(languageService, wpfTextView, adapterFactory, commandFactory)
         {
             _wpfTextView = wpfTextView;
             _commandFactory = commandFactory;
+            _featureServiceFactory = featureServiceFactory;
+        }
+
+        internal void EnableCompletion()
+        {
+            var featureService = _featureServiceFactory.GetOrCreate(WpfTextView);
+            if (!featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
+            {
+                var completionDisabledToken = _completionDisabledToken;
+                if (completionDisabledToken == null)
+                {
+                    return;
+                }
+
+                completionDisabledToken.Dispose();
+                _completionDisabledToken = null;
+
+                var document = _context.ContextBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return;
+                }
+
+                _context.OpenDocumentAndLinkedDocuments(document);
```

Comment:
How is this working? If GetOpenDocumentInCurentContextWithChanges returned a document, it implies this document is *already* open...

Reply:
It seems that `GetOpenDocumentInCurrentContextWithChanges` does not open the document. Not?

Reply:
It does not. But if the document isn't open, it'll return null and you would have taken that early return above. It seems like this code is trying to open a document that is already open based on the earlier check.

Reply:
Thank you! Verified. It seems to be unnecessary. Removing.

---
In reply to: [270161524](https://github.com/dotnet/roslyn/pull/32631#discussion_r270161524) [](ancestors = 270161524)